roachtest: pgurl for nodes X got Y

[(roachtest).decommission/nodes=4/duration=1h0m0s failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1923013&tab=buildLog) on [release-20.1@c856738efa45671ffc6ee31aea7817968e93bb68](https://github.com/cockroachdb/cockroach/commits/c856738efa45671ffc6ee31aea7817968e93bb68):

```
The test failed on branch=release-20.1, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/decommission/nodes=4/duration=1h0m0s/run_1
	cluster.go:2138,cluster.go:2222,decommission.go:218,errgroup.go:57: pgurl for nodes [] empty: :4 from
		stdout:
		
		stderr:

```

<details><summary>More</summary><p>


Artifacts: [/decommission/nodes=4/duration=1h0m0s](https://teamcity.cockroachdb.com/viewLog.html?buildId=1923013&tab=artifacts#/decommission/nodes=4/duration=1h0m0s)
Related:
- #47738 roachtest: decommission/nodes=4/duration=1h0m0s failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-19.2](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-19.2) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Adecommission%2Fnodes%3D4%2Fduration%3D1h0m0s.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

I'm running roachtest-on-Pebble and periodically seeing this error:

```
--- FAIL: acceptance/gossip/locality-address (33.77s)
test artifacts and logs in: artifacts/acceptance/gossip/locality-address/run_1
        cluster.go:2138,cluster.go:2241,gossip.go:487,acceptance.go:94,test_runner.go:753: pgurl for nodes [] empty: :4 from
                stdout:

                stderr:
```

Any thoughts on additional debugging we can add to track down what is happening?

--- a/pkg/cmd/roachtest/cluster.go
+++ b/pkg/cmd/roachtest/cluster.go
@@ -341,10 +341,6 @@ type cmdRes struct {
 // Also note that if the command exits with an error code, its output is also
 // included in cmdRes.err.
 func execCmdEx(ctx context.Context, l *logger, args ...string) cmdRes {
-	// NB: It is important that this waitgroup Waits after cancel() below.
-	var wg sync.WaitGroup
-	defer wg.Wait()
-
 	var cancel func()
 	ctx, cancel = context.WithCancel(ctx)
 	defer cancel()
@@ -356,29 +352,75 @@ func execCmdEx(ctx context.Context, l *logger, args ...string) cmdRes {
 	debugStderrBuffer, _ := circbuf.NewBuffer(1024)
 
 	// Do a dance around https://github.com/golang/go/issues/23019.
-	// Briefly put, passing os.Std{out,err} to subprocesses isn't great for
-	// context cancellation as Run() will wait for any subprocesses to finish.
-	// For example, "roachprod run x -- sleep 20" would wait 20 seconds, even
-	// if the context got canceled right away. Work around the problem by passing
-	// pipes to the command on which we set aggressive deadlines once the context
-	// expires.
+	// When the command we run launches a subprocess, that subprocess receives
+	// a copy of our Command's Stdout/Stderr file descriptor, which effectively
+	// means that the file descriptors close only when that subcommand returns.
+	// However, proactively killing the subcommand is not really possible - we
+	// will only manage to kill the parent process that we launched directly.
+	// In practice this means that if we try to react to context cancellation,
+	// the pipes we read the output from will wait for the *subprocess* to
+	// terminate, leaving us hanging, potentially indefinitely.
+	// To work around it, use pipes and set a read deadline on our (read) end of
+	// the pipes when we detect a context cancellation.
+	//
+	// See TestExecCmd for a test.
+	var closePipes func(ctx context.Context)
+	var wg sync.WaitGroup
 	{
-		rOut, wOut, err := os.Pipe()
-		if err != nil {
-			return cmdRes{err: err}
-		}
-		defer rOut.Close()
-		defer wOut.Close()
 
-		rErr, wErr, err := os.Pipe()
+		var wOut, wErr, rOut, rErr *os.File
+		var cwOnce sync.Once
+		closePipes = func(ctx context.Context) {
+			// Idempotently closes the writing end of the pipes. This is called either
+			// when the process returns or when it was killed due to context
+			// cancellation. In the former case, close the writing ends of the pipe
+			// so that the copy goroutines started below return (without missing any
+			// output). In the context cancellation case, we set a deadline to force
+			// the goroutines to quit eagerly. This is important since the command
+			// may have duplicated wOut and wErr to its possible subprocesses, which
+			// may continue to run for long periods of time, and would otherwise
+			// block this command. In theory this is possible also when the command
+			// returns on its own accord, so we set a (more lenient) deadline in the
+			// first case as well.
+			//
+			// NB: there's also the option (at least on *nix) to use a process group,
+			// but it doesn't look portable:
+			// https://medium.com/@felixge/killing-a-child-process-and-all-of-its-children-in-go-54079af94773
+			cwOnce.Do(func() {
+				if wOut != nil {
+					_ = wOut.Close()
+				}
+				if wErr != nil {
+					_ = wErr.Close()
+				}
+				dur := 10 * time.Second // wait up to 10s for subprocesses
+				if ctx.Err() != nil {
+					dur = 10 * time.Millisecond
+				}
+				deadline := timeutil.Now().Add(dur)
+				if rOut != nil {
+					_ = rOut.SetReadDeadline(deadline)
+				}
+				if rErr != nil {
+					_ = rErr.SetReadDeadline(deadline)
+				}
+			})
+		}
+		defer closePipes(ctx)
+
+		var err error
+		rOut, wOut, err = os.Pipe()
 		if err != nil {
 			return cmdRes{err: err}
 		}
-		defer rErr.Close()
-		defer wErr.Close()
+
+		rErr, wErr, err = os.Pipe()
+		if err != nil {
+			return cmdRes{err: err}
+		}
 
 		cmd.Stdout = wOut
-		wg.Add(3)
+		wg.Add(1)
 		go func() {
 			defer wg.Done()
 			_, _ = io.Copy(l.stdout, io.TeeReader(rOut, debugStdoutBuffer))
@@ -387,29 +429,21 @@ func execCmdEx(ctx context.Context, l *logger, args ...string) cmdRes {
 		if l.stderr == l.stdout {
 			// If l.stderr == l.stdout, we use only one pipe to avoid
 			// duplicating everything.
-			wg.Done()
 			cmd.Stderr = wOut
 		} else {
 			cmd.Stderr = wErr
+			wg.Add(1)
 			go func() {
 				defer wg.Done()
 				_, _ = io.Copy(l.stderr, io.TeeReader(rErr, debugStderrBuffer))
 			}()
 		}
-
-		go func() {
-			defer wg.Done()
-			<-ctx.Done()
-			// NB: setting a more aggressive deadline here makes TestClusterMonitor flaky.
-			now := timeutil.Now().Add(3 * time.Second)
-			_ = rOut.SetDeadline(now)
-			_ = wOut.SetDeadline(now)
-			_ = rErr.SetDeadline(now)
-			_ = wErr.SetDeadline(now)
-		}()
 	}
 
 	err := cmd.Run()
+	closePipes(ctx)
+	wg.Wait()
+
 	if err != nil {
 		// Context errors opaquely appear as "signal killed" when manifested.
 		// We surface this error explicitly.
@@ -417,10 +451,6 @@ func execCmdEx(ctx context.Context, l *logger, args ...string) cmdRes {
 			err = errors.CombineErrors(ctx.Err(), err)
 		}
 
-		// Synchronize access to ring buffers before using them to create an
-		// error to return.
-		cancel()
-		wg.Wait()
 		if err != nil {
 			err = &withCommandDetails{
 				cause:  err,
@@ -430,6 +460,7 @@ func execCmdEx(ctx context.Context, l *logger, args ...string) cmdRes {
 			}
 		}
 	}
+
 	return cmdRes{
 		err:    err,
 		stdout: debugStdoutBuffer.String(),

--- a/pkg/cmd/roachtest/cluster_test.go
+++ b/pkg/cmd/roachtest/cluster_test.go
@@ -20,6 +20,7 @@ import (
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/testutils"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
@@ -66,12 +67,62 @@ func (t testWrapper) logger() *logger {
 	return nil
 }
 
+func TestExecCmd(t *testing.T) {
+	cfg := &loggerConfig{stdout: os.Stdout, stderr: os.Stderr}
+	logger, err := cfg.newLogger("" /* path */)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Run(`success`, func(t *testing.T) {
+		res := execCmdEx(context.Background(), logger, "/bin/bash", "-c", "echo guacamole")
+		require.NoError(t, res.err)
+		require.Contains(t, res.stdout, "guacamole")
+	})
+
+	t.Run(`error`, func(t *testing.T) {
+		res := execCmdEx(context.Background(), logger, "/bin/bash", "-c", "echo burrito; false")
+		require.Error(t, res.err)
+		require.Contains(t, res.stdout, "burrito")
+	})
+
+	t.Run(`returns-on-cancel`, func(t *testing.T) {
+		ctx, cancel := context.WithCancel(context.Background())
+		go func() {
+			time.Sleep(10 * time.Millisecond)
+			cancel()
+		}()
+		tBegin := timeutil.Now()
+		require.Error(t, execCmd(ctx, logger, "/bin/bash", "-c", "sleep 100"))
+		if max, act := 99*time.Second, timeutil.Since(tBegin); max < act {
+			t.Fatalf("took %s despite cancellation", act)
+		}
+	})
+
+	t.Run(`returns-on-cancel-subprocess`, func(t *testing.T) {
+		// The tricky version of the preceding test. The difference is that the process
+		// spawns a stalling subprocess and then waits for it. See execCmdEx for a
+		// detailed discussion of how this is made work.
+		ctx, cancel := context.WithCancel(context.Background())
+		go func() {
+			time.Sleep(10 * time.Millisecond)
+			cancel()
+		}()
+		tBegin := timeutil.Now()
+		require.Error(t, execCmd(ctx, logger, "/bin/bash", "-c", "sleep 100& wait"))
+		if max, act := 99*time.Second, timeutil.Since(tBegin); max < act {
+			t.Fatalf("took %s despite cancellation", act)
+		}
+	})
+}
+
 func TestClusterMonitor(t *testing.T) {
 	cfg := &loggerConfig{stdout: os.Stdout, stderr: os.Stderr}
 	logger, err := cfg.newLogger("" /* path */)
 	if err != nil {
 		t.Fatal(err)
 	}
+
 	t.Run(`success`, func(t *testing.T) {
 		c := &cluster{t: testWrapper{t}, l: logger}
 		m := newMonitor(context.Background(), c)

